File: python_implementation/preprocessing.py
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
import string

# Download necessary NLTK data
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('omw-1.4')

def preprocess_text(text):
    """
    Preprocesses text: lowercasing, tokenization, stopword removal, and lemmatization.
    """
    if not text:
        return ""

    # Lowercasing
    text = text.lower()

    # Tokenization
    tokens = word_tokenize(text)

    # Stopword removal and Punctuation removal
    stop_words = set(stopwords.words('english'))
    tokens = [token for token in tokens if token not in stop_words and token not in string.punctuation]

    # Lemmatization
    lemmatizer = WordNetLemmatizer()
    tokens = [lemmatizer.lemmatize(token) for token in tokens]

    return " ".join(tokens)

File: src/services/preprocessing.ts
import natural from 'natural';

const tokenizer = new natural.WordTokenizer();
const stemmer = natural.PorterStemmer;
const stopwords = new Set(natural.stopwords);

/**
 * Preprocesses text: lowercasing, tokenization, stopword removal, and stemming (as a proxy for lemmatization in JS).
 */
export function preprocessText(text: string): string {
  if (!text) return '';

  // Lowercasing
  const lowerText = text.toLowerCase();

  // Tokenization
  const tokens = tokenizer.tokenize(lowerText) || [];

  // Stopword removal and Stemming
  const processedTokens = tokens
    .filter(token => !stopwords.has(token) && token.length > 1)
    .map(token => stemmer.stem(token));

  return processedTokens.join(' ');
}
